Create New Style Command Button

DoctorRyan opened the pull request:
This PR creates a new button in the pallette editor that creates a new style.  Button is large and easy access for a faster and easier workflow.  Original code developed by Turtletooth for Tahoma.  Given permission to develop within Openoonz.

Co-Authored-By: Jeremy Bullock <turtletooth@users.noreply.github.com>

DoctorRyan commented:
Okay, so I deleted my last commit because it was so messy.  This commit is for the new style command for the palette editor within Opentoonz.  @shun-iwasawa @RodneyBaker @manongjohn please review for submission to the source code.  Thank you.

manongjohn commented:
There is no need to keep closing and reopening new PRs every time you make changes to your working branch. Now the prior discussion of this change is no longer attached to this PR. There are ways to revise your commit history on the working branch. We also reserve the right to Merge-Squash your changes so it appears as 1 commit in the master.

Learn to use `git reset HEAD~# --soft` (#=commits to basically revert but keep changes), make changes and new commits as needed and then `git push --force` to rewrite the commit history of the branch. 

Try not to overuse this. If people pull this branch into theirs and you rewrite history, it can be problematic for those that need to repull your branch for new changes.

While there is nothing wrong with having multiple commits especially when fixing a bug found after pushing, learn to stage your sets of changes across multiple files instead of committing every little change to each file. Then make 1 commit when you think you are done to help minimize your commits.

DoctorRyan commented:
I apologize.  Thought I was making it easier and not harder.  I learned another lesson I guess.  I did learn about how to not commit a bunch of times making that look better.  Didn't know you can combine the commits together either.  Hope you are doing well @manongjohn.  Thanks for the advice.

DoctorRyan commented:
Thank you for the review. I will work on those changes.

On Thu, Jul 9, 2020, 1:16 PM shun-iwasawa <notifications@github.com> wrote:

> *@shun-iwasawa* requested changes on this pull request.
>
> Thank you @DoctorRyan <https://github.com/DoctorRyan> for porting the
> feature. And I didn't know the Co-Authored-By: commit trailer. It's very
> nice.
> ------------------------------
>
> In toonz/sources/toonzqt/paletteviewergui.cpp
> <https://github.com/opentoonz/opentoonz/pull/3394#discussion_r452350342>:
>
> > @@ -809,12 +829,12 @@ void PageViewer::paintEvent(QPaintEvent *e) {
>          }
>        }
>
> -      // draw the frame border if the style is selected or current
> -      if (m_styleSelection->isSelected(m_page->getIndex(), i) ||
> -          currentStyleIndex == styleIndex) {
> -        p.setBrush(Qt::NoBrush);
> -        p.drawRect(chipRect.adjusted(3, 3, -3, -3));
> -      }
> +      //// draw the frame border if the style is selected or current
> +      //if (m_styleSelection->isSelected(m_page->getIndex(), i) ||
> +      //    currentStyleIndex == styleIndex) {
> +      //  p.setBrush(Qt::NoBrush);
> +      //  p.drawRect(chipRect.adjusted(3, 3, -3, -3));
> +      //}
>
> Here you have deleted the inner-frame indicating the selected & current
> styles.
> I think some users (at least staffs in our studio) may prefer to keep it
> to make the selected styles prominent.
>
> Can you please separate this change into another PR since it seems to be
> outside the scope of the proposed feature?
> ------------------------------
>
> In toonz/sources/toonzqt/paletteviewergui.cpp
> <https://github.com/opentoonz/opentoonz/pull/3394#discussion_r452359976>:
>
> > @@ -1325,14 +1367,11 @@ bool PageViewer::event(QEvent *e) {
>        TColorStyle *style = m_page->getStyle(indexInPage);
>        if (style) {
>          int styleIndex = m_page->getStyleId(indexInPage);
> -        toolTip        = "#" + QString::number(styleIndex) + " " +
> -                  QString::fromStdWString(style->getName());
> -
> -        int shortcutKey = m_page->getPalette()->getStyleShortcut(styleIndex);
> -        if (shortcutKey > 0)
> -          toolTip += QString::fromStdWString(std::wstring(L" (") +
> -                                             (wchar_t)shortcutKey + L")");
>
> Here you are removing the style tool tip.
> Again, I think some users may need to keep this because it is the only way
> to know the style names in the small thumbnails view and also to see the
> entire name which is too long to display in the thumbnails.
>
> Can you please separate this change into another PR too ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opentoonz/opentoonz/pull/3394#pullrequestreview-445790045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APTY7C4MIUIKHG2L3CSTZ6DR2X3PTANCNFSM4OVGW4ZQ>
> .
>

DoctorRyan commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -809,12 +829,12 @@ void PageViewer::paintEvent(QPaintEvent *e) {
+      //    currentStyleIndex == styleIndex) {
+      //  p.setBrush(Qt::NoBrush);
+      //  p.drawRect(chipRect.adjusted(3, 3, -3, -3));
+      //}
```
I went ahead and set this back to normal.

DoctorRyan commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -1325,14 +1367,11 @@ bool PageViewer::event(QEvent *e) {
-        int shortcutKey = m_page->getPalette()->getStyleShortcut(styleIndex);
-        if (shortcutKey > 0)
-          toolTip += QString::fromStdWString(std::wstring(L" (") +
-                                             (wchar_t)shortcutKey + L")");
```
Also reverted this back to the original code.  I can do PR's on the last two changes if you wish @shun-iwasawa but it will be a little while unless you deem them necessary.  Thanks.

yokobell commented:
I wonder if this can be made optional? It stands out way too much in large thumbnail view, that it is actually bigger than the swatches themselves and has a different design, just the plus symbol without any box around it would be better imho?

![image](https://user-images.githubusercontent.com/68205893/87300741-4f5a1200-c506-11ea-9114-3e25a239b1e9.png)

I would prefer to hide it though... since the button changes place all the time I just don't see how it is anymore intuitive than the current button, I fear I would hit it by accident all the time.

In list view there is some dead space left over where the button would be when you lock the palette, which is not a big deal. But the plus button being white on light theme is not good, it should be same as text color imho...

![image](https://user-images.githubusercontent.com/68205893/87302171-ce504a00-c508-11ea-964a-d71e5cc7c355.png)

artisteacher commented:
In Tahoma, the small thumbnails are the default & the button works beautifully. The shifting location hasn't been a concern - I've never hit it by accident. In OT, I most often use the context menu command to make new styles since it's faster than moving the cursor up to the button. I actually use the button in Tahoma - for me, it is much more intuitive.

The button in the light theme could definitely use much better contrast.

DoctorRyan commented:
@yokobell  Thank you for your suggestions.  I can see about making some changes to the big button design.  As @artisteacher pointed out in Tahoma the chips are small by default making this look a lot better.  Also, I had the button design cleaned up a little but @shun-iwasawa requested that not be included in this PR.  I can try out your suggestions though for his approval after he reviews this PR.

shun-iwasawa commented:
@yokobell 

>I would prefer to hide it though... since the button changes place all the time I just don't see how it is anymore intuitive than the current button, I fear I would hit it by accident all the time.

I have the same concern and I'm thinking of making the button optional, putting some option in the hamburger menu of the palette toolbar.

I'll merge this since the appearance can be modified in the future development.

Thank you @DoctorRyan for porting the useful feature. LGTM

shun-iwasawa commented:
Let me call Jenkins before merging.

DoctorRyan commented:
Thanks.  Did you have any concerns with the size of the button before you
merge?  I can maybe make the rectangle smaller if needed?

On Mon, Jul 13, 2020, 1:39 PM shun-iwasawa <notifications@github.com> wrote:

> Let me call Jenkins before merging.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opentoonz/opentoonz/pull/3394#issuecomment-657695636>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APTY7C5H5BYGEWKMCHAVZ6LR3NBELANCNFSM4OVGW4ZQ>
> .
>

shun-iwasawa commented:
@DoctorRyan 

>Did you have any concerns with the size of the button before you merge?  I can maybe make the rectangle smaller if needed?

I think it's OK to merge this as-is since the size of the button can be adjusted in the future development.
Of course it's also great if you could and would like to include the adjustment in this PR as well.
If so, please post a comment or add `[WIP]` (work in progress) to the top of the PR title so that reviewers notice your intention.

DoctorRyan commented:
I think we can go ahead and merge for now.  But i will come back to this at
a later date to bring it in line with the other chips.  Thank you
@shun-iwasawa.
On Mon, Jul 13, 2020, 9:49 PM shun-iwasawa <notifications@github.com> wrote:

> @DoctorRyan <https://github.com/DoctorRyan>
>
> Did you have any concerns with the size of the button before you merge? I
> can maybe make the rectangle smaller if needed?
>
> I think it's OK to merge this as-is since the size of the button can be
> adjusted in the future development.
> Of course it's also great if you could and would like to include the
> adjustment in this PR as well.
> If so, please post a comment or add [WIP] (work in progress) to the top
> of the PR title so that reviewers notice your intention.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opentoonz/opentoonz/pull/3394#issuecomment-657918682>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APTY7C3CPLHMYCXLO7HYVA3R3O2T7ANCNFSM4OVGW4ZQ>
> .
>
